### DISCUSSION TITLE
4.0.0 Beta 1 - The King is dead, long live the King

### DISCUSSION THREAD

**Discussion by Kaspik** [MessageKit/MessageKit#1711] [Announcements]
iOS 12 has been around for a while, but with upcoming WWDC, we are ready to let this version go and rest in peace forever. The same could be said about CocoaPods - what used to be the major and most important thing on iOS for developing apps with 3rd party libraries - has been replaced by natively supported way called Swift Package Manager.

**Sit tibi terra levis.**

***

See [MIGRATION_GUIDE.md](https://github.com/MessageKit/MessageKit/blob/main/Documentation/MIGRATION_GUIDE.md) for migration to the new V4.

### Changed
   
- **Breaking change**: Dropped CocoaPods support
- **Breaking change**: Dropped support for iOS 12 [2bd234b](https://github.com/MessageKit/MessageKit/commit/2bd234b1e878f392089f166d6868ce644d6c9e95) by [@martinpucik](https://github.com/martinpucik)
- **Breaking change**: Moved messageInputBar from inputAccessoryView to a subview in MessagesViewController [#1704](https://github.com/MessageKit/MessageKit/pull/1704) by [@martinpucik](https://github.com/martinpucik)
- **Deprecation**: Deprecated `maintainPositionOnKeyboardFrameChangedMoved` in favor of `maintainPositionOnInputBarHeightChanged` which better describes the intended use of this property [#1704](https://github.com/MessageKit/MessageKit/pull/1705) by [@martinpucik](https://github.com/martinpucik)
- **Breaking change**: Added an argument to `messageContainerMaxWidth` [cd4f75b](https://github.com/MessageKit/MessageKit/commit/cd4f75b561129fc25e6c4576000e5a92ccd81cad) by [@martinpucik](https://github.com/martinpucik)
    ```swift
    MessageSizeCalculator.messageContainerMaxWidth(for message: MessageType) -> CGFloat
    ```
    now has IndexPath argument
    ```swift
    MessageSizeCalculator.messageContainerMaxWidth(for message: MessageType, at indexPath: IndexPath) -> CGFloat
    ```
- **Breaking change**: Added an argument to `messageContainerSize` [cd4f75b](https://github.com/MessageKit/MessageKit/commit/cd4f75b561129fc25e6c4576000e5a92ccd81cad) by [@martinpucik](https://github.com/martinpucik)
    ```swift
    MessageSizeCalculator.messageContainerSize(for message: MessageType) -> CGSize
    ```
    now has IndexPath argument 
    ```swift
    MessageSizeCalculator.messageContainerSize(for message: MessageType, at indexPath: IndexPath) -> CGSize
    ```

### Added

- New method in `MessagesLayoutDelegate` for setting message avatar size [ddfc814](https://github.com/MessageKit/MessageKit/commit/ddfc814d325ee5aa238484c90128d32e5a72a49b) by [@martinpucik](https://github.com/martinpucik)
- `MessageInputBarKind` enum for customizing `messageInputBar` inside `inputContainerView` [#1707](https://github.com/MessageKit/MessageKit/pull/1707) by [@martinpucik](https://github.com/martinpucik)

### Removed

- NSConstraintLayoutSet.swift [#1700](https://github.com/MessageKit/MessageKit/pull/1700) by [@martinpucik](https://github.com/martinpucik) 

<hr /><em>This discussion was created from the release <a href='https://github.com/MessageKit/MessageKit/releases/tag/4.0.0.beta_1'>4.0.0 Beta 1 - The King is dead, long live the King</a>.</em>

**Comment by erwanmace** [comment]
Hi, any visibility on the progress of 4.0 branch? Estimated completion?
Any migration guide we can use with beta 2?

**Reply by Kaspik** [reply]
There should be no migration between beta 1 and beta 2.

We hope we will get 4.0.0 out in June. :)

**Reply by erwanmace** [reply]
Thanks @Kaspik 
I meant migration from 3.x to 4.0...

**Reply by Kaspik** [reply]
Yeah check the migration guide. ✌🏼

**Reply by erwanmace** [reply]
Using 4.0 beta 2 in my app... works great. Thank you guys.
Any update on final release?

**Reply by Kaspik** [reply]
This weekend probably. 🤞🏼
